core/src/main/java/org/opensearch/sql/planner/streaming/watermark/BoundedOutOfOrderWatermarkGenerator.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.planner.streaming.watermark;
+
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Watermark generator that generates watermark with bounded out-of-order delay.
+ */
+@RequiredArgsConstructor
+public class BoundedOutOfOrderWatermarkGenerator implements WatermarkGenerator {
+
+  /** The maximum out-of-order allowed in millisecond. */
+  private final long maxOutOfOrderAllowed;
+
+  /** The maximum timestamp seen so far in millisecond. */
+  private long maxTimestamp;
+
+  @Override
+  public long generate(long timestamp) {
```

Comment:
implementation is not threadsafe? leave some comments for future improvement. WatermarkGenerator object will be used across operator?

Reply:
Specific WatermarkGenerator will be used by a new WatermarkGenerateOperator which is not shared across different query plan or source.